File: url_test.mjs
import { strict as assert } from "node:assert";
import { dial } from "./client.mjs";

const originalWebSocket = globalThis.WebSocket;
const originalFetch = globalThis.fetch;

class RecordingWebSocket {
  static urls = [];

  constructor(url) {
    RecordingWebSocket.urls.push(String(url));
    queueMicrotask(() => this.onopen?.());
  }

  send() {}

  close() {
    this.onclose?.();
  }
}

try {
  globalThis.WebSocket = RecordingWebSocket;

  const cases = [
    ["root endpoint", "http://example.com", "ws://example.com/"],
    [
      "nested path and trailing slash",
      "http://example.com/one/two/",
      "ws://example.com/one/two/",
    ],
    [
      "existing query",
      "http://example.com/wd/?mode=fast",
      "ws://example.com/wd/?mode=fast",
    ],
    [
      "encoded query value",
      "http://example.com/wd/?next=https%3A%2F%2Fupstream.example%2Fa%2Fb",
      "ws://example.com/wd/?next=https%3A%2F%2Fupstream.example%2Fa%2Fb",
    ],
    [
      "literal URL in query",
      "http://example.com/wd/?next=https://upstream.example/a/",
      "ws://example.com/wd/?next=https://upstream.example/a/",
    ],
    [
      "HTTPS with HTTP query value",
      "https://example.com/wd/?next=http://upstream.example/a/",
      "wss://example.com/wd/?next=http://upstream.example/a/",
    ],
    [
      "IPv6 host",
      "http://[2001:db8::1]:8080/wd/",
      "ws://[2001:db8::1]:8080/wd/",
    ],
    [
      "fragment",
      "http://example.com/wd/?x=1#next=https://fragment.example/",
      "ws://example.com/wd/?x=1#next=https://fragment.example/",
    ],
  ];

  for (const [name, endpoint, expected] of cases) {
    RecordingWebSocket.urls = [];
    const conn = await dial(endpoint, {
      transport: "ws",
      pingIntervalMs: 60_000,
    });
    assert.equal(RecordingWebSocket.urls[0], expected, name);
    assert.equal(conn.url, endpoint, `${name}: public URL changed`);
    await conn.close();
  }

  const calls = [];
  globalThis.fetch = async (resource, init = {}) => {
    calls.push({ url: String(resource), init });
    if (init.method === "POST") return new Response(null, { status: 204 });
    const body = new ReadableStream({
      start(controller) {
        controller.enqueue(
          new TextEncoder().encode("event: sid\ndata: session/id\n\n"),
        );
      },
    });
    return new Response(body, { status: 200 });
  };

  const endpoint =
    "http://example.com/wd/?next=https://upstream.example/a%2Fb&token=a%20b#section";
  const postURL =
    "http://example.com/wd/?next=https://upstream.example/a%2Fb&token=a%20b&s=session%2Fid#section";
  const conn = await dial(endpoint, {
    transport: "sse",
    pingIntervalMs: 60_000,
  });
  assert.equal(conn.url, endpoint);
  assert.equal(calls[0].url, endpoint, "SSE endpoint changed");

  await conn.write("hello");
  assert.equal(calls[1].url, postURL, "session parameter URL is malformed");

  conn.ping();
  await Promise.resolve();
  const pingURL = new URL(calls[2].url);
  assert.equal(pingURL.hash, "#section");
  assert.ok(pingURL.searchParams.has("ping"));
  assert.ok(
    calls[2].url.startsWith(postURL.slice(0, -"#section".length) + "&ping="),
    "ping parameter did not preserve the existing raw query",
  );

  await conn.close();
  assert.equal(
    calls[3].url,
    postURL.slice(0, -"#section".length) + "&close=1#section",
    "close parameter URL is malformed",
  );

  console.log("all URL tests passed");
} finally {
  globalThis.WebSocket = originalWebSocket;
  globalThis.fetch = originalFetch;
}

File: wt/url_test.go
package wt

import (
	"testing"

	"github.com/stretchr/testify/require"
)

func TestWebtransportURL(t *testing.T) {
	tests := []struct {
		name string
		in   string
		want string
	}{
		{
			name: "http is promoted",
			in:   "http://example.com",
			want: "https://example.com",
		},
		{
			name: "https passes through",
			in:   "https://example.com",
			want: "https://example.com",
		},
		{
			name: "nested path and trailing slash",
			in:   "http://example.com/one/two/",
			want: "https://example.com/one/two/",
		},
		{
			name: "existing query",
			in:   "https://example.com/wd/?mode=fast",
			want: "https://example.com/wd/?mode=fast",
		},
		{
			name: "encoded query value",
			in:   "https://example.com/wd/?next=https%3A%2F%2Fupstream.example%2Fa%2Fb",
			want: "https://example.com/wd/?next=https%3A%2F%2Fupstream.example%2Fa%2Fb",
		},
		{
			name: "literal URL in query",
			in:   "https://example.com/wd/?next=https://upstream.example/a/",
			want: "https://example.com/wd/?next=https://upstream.example/a/",
		},
		{
			name: "http query value is left alone",
			in:   "https://example.com/wd/?next=http://upstream.example/a/",
			want: "https://example.com/wd/?next=http://upstream.example/a/",
		},
		{
			name: "IPv6 host",
			in:   "http://[2001:db8::1]:8080/wd/",
			want: "https://[2001:db8::1]:8080/wd/",
		},
		{
			name: "fragment",
			in:   "https://example.com/wd/?x=1#next=https://fragment.example/",
			want: "https://example.com/wd/?x=1#next=https://fragment.example/",
		},
		{
			name: "uppercase scheme",
			in:   "HTTP://example.com/wd/",
			want: "https://example.com/wd/",
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			got, err := webtransportURL(tc.in)
			require.NoError(t, err)
			require.Equal(t, tc.want, got)
		})
	}
}

// TestWebtransportURLRejectsScheme pins the deliberate divergence from
// websocketURL, which passes unknown schemes through untouched. WebTransport
// has no cleartext or ws form, and failing here beats failing inside the dial.
func TestWebtransportURLRejectsScheme(t *testing.T) {
	for _, in := range []string{
		"ws://example.com/wd/",
		"wss://example.com/wd/",
		"ftp://example.com/wd/",
		"example.com/wd/", // no scheme at all
	} {
		t.Run(in, func(t *testing.T) {
			_, err := webtransportURL(in)
			require.Error(t, err)
			require.Contains(t, err.Error(), "webtransport requires an https URL")
		})
	}
}

func TestWebtransportURLRejectsUnparseable(t *testing.T) {
	_, err := webtransportURL("http://[::1")
	require.Error(t, err)
	require.Contains(t, err.Error(), "parse endpoint URL")
}
